---
title: Testing
permalink: technical-docs/testing
parent: Using PharmCAT
nav_order: 5
---
# Autogenerated VCF Testing

TL;DR:

```commandline
# make exactVcfTests
```

## Generating VCFs to Test

`/src/scripts/vcf_generator/test_gen.py` is responsible for generating VCF files used to test the `NamedAlleleMatcher`.  

It relies on the allele definitions to generate all possible combinations of variations that will lead to specific diplotype calls for a single gene.

`/src/scripts/vcf_generator/generate_vcf_test_data.sh` iterates through all genes, calling either:

* `/src/scripts/vcf_generator/test_gen.py`
* or `/src/scripts/vcf_generator/test_gen_missing.py` (will also generate VCFs with missing positions) if the `-m` flag is provided

and storing the VCFs in `/build/testVcf`.


## Testing NamedAlleleMatcher

These VCFs are then run through the `NamedAlleleMatcher` using `org.pharmgkb.pharmcat.haplotype.AutogeneratedVcfTester`:

* By default, test will pass if expected result is one of the highest matching results from the `NamedAlleleMatcher`.
* In fuzzy-match mode, test will pass if expected result matches any result from the `NamedAlleleMatcher`.
* In exact-match mode, test will only pass if the `NamedAlleleMatcher` produces a single result that matches the expected result.

The "expected result" is determined based on the filename of the input VCF.

Results are saved to `/build/autogeneratedTestResults`.  A summary is saved to `/build/autogeneratedTestResults/autogenerated_test_report.txt`.  The directory will also contain a copy of the VCF and `NamedAlleleMatcher`'s results for each failing test.


## Make Targets

To simplify testing, use one of these make targets:

* vcfTests
* vcfMissingTests
* exactVcfTests
* exactVcfMissingTests
* fuzzyVcfTests
* fuzzyVcfMissingTests

For example:

```commandline
# make exactVcfTests
# make exactVcfMissingTests
```
